File: lob_replay.cpp
// lob_replay.cpp — Portable, Arrow‑20‑ready Limit‑Order‑Book replay engine
// ======================================================================
// 
// High-performance limit order book replay engine that processes market data
// from Parquet files and provides detailed latency metrics for each message.
//
// FEATURES:
// * Zero-copy Arrow columnar data processing
// * Real limit order book semantics (add/cancel/execute/replace)
// * Sub-10ns per-message latency measurement using cycle counters
// * Comprehensive throughput and latency statistics (p50/p95/p99/max)
// * Support for BX/ITCH-like market data schemas
//
// SETUP & USAGE:
// ======================================================================
//
// 1. Install Dependencies (macOS with Homebrew)
// ----------------------------------------------------------------------
//    brew install apache-arrow
//    # For other platforms, install Apache Arrow 20+ via your package manager
//
// 2. Download Sample Data
// ----------------------------------------------------------------------
//    wget https://emi.nasdaq.com/ITCH/Nasdaq%20BX%20ITCH/20190530.BX_ITCH_50.gz
//
// 3. Convert ITCH to Parquet (requires itch_to_parquet.py)
// ----------------------------------------------------------------------
//    python data_pipeline/itch_to_parquet.py 20190530.BX_ITCH_50.gz bx_20190530.parquet
//
// 4. Build the Executable
// ----------------------------------------------------------------------
//    INC=$(brew --prefix)/include
//    LIB=$(brew --prefix)/lib
//    g++ -O3 -march=native -std=c++20 lob_replay.cpp \
//        -I${INC} -L${LIB} -larrow -lparquet -pthread -o lob_replay
//
//    # If runtime loader can't find dylibs:
//    export DYLD_LIBRARY_PATH="${LIB}:$DYLD_LIBRARY_PATH"
//
// 5. Run the Replay Engine
// ----------------------------------------------------------------------
//    ./lob_replay bx_20190530.parquet
//
// EXPECTED PARQUET SCHEMA:
// ======================================================================
// Column names (all non-nullable):
//   ts:uint64        - timestamp
//   oid:uint64       - order ID
//   side:uint8       - 0=bid, 1=ask
//   px:uint32        - price
//   qty:uint32       - quantity
//   m:string         - message type ("A"=add, "C"=cancel, "E"=execute, "U"=replace)
//   stock:string     - symbol
//   new_oid:uint64   - new order ID (for replace messages)
//   new_px:uint32    - new price (for replace messages)
//   new_qty:uint32   - new quantity (for replace messages)
//
// PERFORMANCE NOTES:
// ======================================================================
// * Uses __builtin_readcyclecounter() for sub-10ns latency measurement
// * Optimized for Apple Silicon and x86-64 architectures
// * Processes millions of messages per second with detailed per-message metrics
// * Memory-efficient order book implementation using hash maps
//
// ======================================================================

#include <arrow/api.h>
#include <arrow/io/api.h>
#include <parquet/arrow/reader.h>
#include <parquet/exception.h>

#include <chrono>
#include <unordered_map>
#include <vector>
#include <iostream>
#include <cstring>
#include <algorithm>
#include <string_view>

// ──────────────────────────── Order‑book structs ──────────────────────
struct Order {
    uint64_t oid;
    uint32_t px;
    uint32_t qty;
    uint8_t  side;   // 0 = bid, 1 = ask
};

struct OrderBook {
    std::unordered_map<uint64_t, Order> live;           // oid → Order
    std::unordered_map<uint32_t, uint32_t> bid_size;    // px → qty
    std::unordered_map<uint32_t, uint32_t> ask_size;

    inline void add(uint64_t oid, uint8_t side, uint32_t px, uint32_t qty) {
        live.emplace(oid, Order{oid, px, qty, side});
        auto &lvl = (side == 0 ? bid_size : ask_size);
        lvl[px] += qty;
    }
    inline void cancel(uint64_t oid) {
        auto it = live.find(oid);
        if (it == live.end()) return;
        auto &lvl = (it->second.side == 0 ? bid_size : ask_size);
        auto lit = lvl.find(it->second.px);
        if (lit != lvl.end()) {
            lit->second -= it->second.qty;
            if (lit->second == 0) lvl.erase(lit);
        }
        live.erase(it);
    }
    inline void execute(uint64_t oid, uint32_t qty_exec) {
        auto it = live.find(oid);
        if (it == live.end()) return;
        auto &o   = it->second;
        auto &lvl = (o.side == 0 ? bid_size : ask_size);
        auto lit  = lvl.find(o.px);
        uint32_t decr = std::min(qty_exec, o.qty);
        if (lit != lvl.end()) {
            lit->second -= decr;
            if (lit->second == 0) lvl.erase(lit);
        }
        o.qty -= decr;
        if (o.qty == 0) live.erase(it);
    }
    inline void replace(uint64_t oid, uint64_t new_oid, uint32_t new_px, uint32_t new_qty) {
        auto it = live.find(oid);
        if (it == live.end()) return;
        uint8_t side = it->second.side;
        cancel(oid);
        add(new_oid, side, new_px, new_qty);
    }
};

// ───────────────────────── Low‑level cycle timer (ns precision) ─────────────
static inline uint64_t rdtsc() {
    return __builtin_readcyclecounter(); // works on x86‑64 & Apple‑silicon
}

// ─────────────────────────────────── Main Application ─────────────────────────────
int main(int argc, char **argv) {
    if (argc != 2) {
        std::cerr << "Usage: ./lob_replay <bx_YYYYMMDD.parquet>\n";
        return 1;
    }

    // 1. Open Parquet file and initialize Arrow reader ----------------------
    auto maybe_infile = arrow::io::ReadableFile::Open(argv[1]);
    if (!maybe_infile.ok()) { std::cerr << maybe_infile.status() << '\n'; return 1; }
    std::shared_ptr<arrow::io::RandomAccessFile> infile = *maybe_infile;

    parquet::ReaderProperties pq_props(arrow::default_memory_pool());
    std::unique_ptr<parquet::ParquetFileReader> pq_reader = parquet::ParquetFileReader::Open(infile, pq_props);

    parquet::ArrowReaderProperties ar_props;
    std::unique_ptr<parquet::arrow::FileReader> reader;
    PARQUET_THROW_NOT_OK(parquet::arrow::FileReader::Make(arrow::default_memory_pool(), std::move(pq_reader), ar_props, &reader));

    std::shared_ptr<arrow::Table> table;
    PARQUET_THROW_NOT_OK(reader->ReadTable(&table));
    PARQUET_ASSIGN_OR_THROW(table, table->CombineChunks(arrow::default_memory_pool()));

    const int64_t rows = table->num_rows();

    // 2. Extract column arrays (zero-copy access to Arrow data) --------------
    auto ts_col      = table->GetColumnByName("ts")->chunk(0);
    auto oid_col     = table->GetColumnByName("oid")->chunk(0);
    auto side_col    = table->GetColumnByName("side")->chunk(0);
    auto px_col      = table->GetColumnByName("px")->chunk(0);
    auto qty_col     = table->GetColumnByName("qty")->chunk(0);
    auto m_col       = table->GetColumnByName("m")->chunk(0);
    auto stock_col   = table->GetColumnByName("stock")->chunk(0);
    auto new_oid_col = table->GetColumnByName("new_oid")->chunk(0);
    auto new_px_col  = table->GetColumnByName("new_px")->chunk(0);
    auto new_qty_col = table->GetColumnByName("new_qty")->chunk(0);

    auto ts_arr      = std::static_pointer_cast<arrow::UInt64Array>(ts_col);
    auto oid_arr     = std::static_pointer_cast<arrow::UInt64Array>(oid_col);
    auto side_arr    = std::static_pointer_cast<arrow::UInt8Array >(side_col);
    auto px_arr      = std::static_pointer_cast<arrow::UInt32Array>(px_col);
    auto qty_arr     = std::static_pointer_cast<arrow::UInt32Array>(qty_col);
    auto m_arr       = std::static_pointer_cast<arrow::StringArray>(m_col);
    auto stock_arr   = std::static_pointer_cast<arrow::StringArray>(stock_col);
    auto new_oid_arr = std::static_pointer_cast<arrow::UInt64Array>(new_oid_col);
    auto new_px_arr  = std::static_pointer_cast<arrow::UInt32Array>(new_px_col);
    auto new_qty_arr = std::static_pointer_cast<arrow::UInt32Array>(new_qty_col);

    // 3. Initialize per‑symbol order books -----------------------------------
    std::unordered_map<std::string, OrderBook> books;
    books.reserve(256);

    std::vector<uint64_t> latencies; latencies.reserve(rows);

    // 4. Execute order book replay with per-message latency tracking ---------
    const auto wall_t0 = std::chrono::steady_clock::now();
    const uint64_t cyc0 = rdtsc();

    for (int64_t i = 0; i < rows; ++i) {
        uint64_t tic = rdtsc();

        uint64_t oid  = oid_arr->Value(i);
        uint8_t  side = side_arr->Value(i);
        uint32_t px   = px_arr->Value(i);
        uint32_t qty  = qty_arr->Value(i);
        std::string_view m_type = m_arr->GetView(i);     // "A", "C", "E", "U" …
        const std::string &sym  = stock_arr->GetString(i);

        OrderBook &ob = books.emplace(sym, OrderBook{}).first->second;

        if (m_type == "A") {
            ob.add(oid, side, px, qty);
        } else if (m_type == "C") {
            ob.cancel(oid);
        } else if (m_type == "E") {
            ob.execute(oid, qty);
        } else if (m_type == "U") {
            uint64_t new_oid = new_oid_arr->Value(i);
            uint32_t new_px  = new_px_arr->Value(i);
            uint32_t new_qty = new_qty_arr->Value(i);
            ob.replace(oid, new_oid, new_px, new_qty);
        }

        uint64_t toc = rdtsc();
        latencies.push_back(toc - tic);
    }

    const uint64_t cyc1 = rdtsc();
    const auto wall_t1 = std::chrono::steady_clock::now();

    // 5. Calculate and display performance metrics ---------------------------
    double wall_seconds = std::chrono::duration<double>(wall_t1 - wall_t0).count();
    double throughput   = rows / wall_seconds;

    std::sort(latencies.begin(), latencies.end());
    auto pct = [&](double p){ return latencies[static_cast<size_t>(p * latencies.size())]; };

    std::cout << "LOB Replay Metrics\n";
    std::cout << "──────────────────\n";
    std::cout << "Rows processed     : " << rows << '\n';
    std::cout << "Total wall time (s): " << wall_seconds << '\n';
    std::cout << "Throughput (msg/s) : " << throughput/1e6 << " M" << '\n';
    std::cout << "Latency (cycles) — p50 : " << pct(0.50) << '\n';
    std::cout << "Latency (cycles) — p95 : " << pct(0.95) << '\n';
    std::cout << "Latency (cycles) — p99 : " << pct(0.99) << '\n';
    std::cout << "Latency (cycles) — max : " << latencies.back() << '\n';

    return 0;
}

File: lob_replay.py
#!/usr/bin/env python3
"""
lob_replay.py — High-performance Python Limit-Order-Book replay engine
======================================================================

Python implementation of the C++ LOB replay engine with optimizations:
* PyArrow zero-copy columnar data processing
* Numba JIT compilation for hot paths
* Efficient collections.defaultdict for order books
* High-resolution time measurement
* Vectorized operations where possible

SETUP & USAGE:
======================================================================

1. Install Dependencies
----------------------------------------------------------------------
   pip install pyarrow numba pandas numpy

2. Run the Replay Engine
----------------------------------------------------------------------
   python lob_replay.py bx_20190530.parquet

PERFORMANCE OPTIMIZATIONS:
======================================================================
* Numba JIT compilation for order book operations
* PyArrow zero-copy data access
* Efficient hash maps using defaultdict
* Minimal Python overhead in hot loops
* High-resolution timing using time.perf_counter_ns()

======================================================================
"""

import sys
import time
import numpy as np
import pyarrow as pa
import pyarrow.parquet as pq
from collections import defaultdict
from typing import Dict, List, Optional
import numba
from numba import jit, types
from numba.typed import Dict as NumbaDict


# ──────────────────────────── Order-book with Numba JIT ──────────────────────
@numba.experimental.jitclass([
    ('oid', numba.uint64),
    ('px', numba.uint32),
    ('qty', numba.uint32),
    ('side', numba.uint8),
])
class Order:
    def __init__(self, oid, px, qty, side):
        self.oid = oid
        self.px = px
        self.qty = qty
        self.side = side


@jit(nopython=True, cache=True)
def add_order(live_orders, bid_levels, ask_levels, oid, side, px, qty):
    """Add order to book with JIT compilation"""
    live_orders[oid] = (px, qty, side)
    
    if side == 0:  # bid
        if px in bid_levels:
            bid_levels[px] += qty
        else:
            bid_levels[px] = qty
    else:  # ask
        if px in ask_levels:
            ask_levels[px] += qty
        else:
            ask_levels[px] = qty


@jit(nopython=True, cache=True)
def cancel_order(live_orders, bid_levels, ask_levels, oid):
    """Cancel order from book with JIT compilation"""
    if oid not in live_orders:
        return
    
    px, qty, side = live_orders[oid]
    
    if side == 0:  # bid
        if px in bid_levels:
            bid_levels[px] -= qty
            if bid_levels[px] == 0:
                del bid_levels[px]
    else:  # ask
        if px in ask_levels:
            ask_levels[px] -= qty
            if ask_levels[px] == 0:
                del ask_levels[px]
    
    del live_orders[oid]


@jit(nopython=True, cache=True)
def execute_order(live_orders, bid_levels, ask_levels, oid, qty_exec):
    """Execute order with JIT compilation"""
    if oid not in live_orders:
        return
    
    px, qty, side = live_orders[oid]
    decr = min(qty_exec, qty)
    
    if side == 0:  # bid
        if px in bid_levels:
            bid_levels[px] -= decr
            if bid_levels[px] == 0:
                del bid_levels[px]
    else:  # ask
        if px in ask_levels:
            ask_levels[px] -= decr
            if ask_levels[px] == 0:
                del ask_levels[px]
    
    new_qty = qty - decr
    if new_qty == 0:
        del live_orders[oid]
    else:
        live_orders[oid] = (px, new_qty, side)


@jit(nopython=True, cache=True)
def replace_order(live_orders, bid_levels, ask_levels, oid, new_oid, new_px, new_qty):
    """Replace order with JIT compilation"""
    if oid not in live_orders:
        return
    
    _, _, side = live_orders[oid]
    cancel_order(live_orders, bid_levels, ask_levels, oid)
    add_order(live_orders, bid_levels, ask_levels, new_oid, side, new_px, new_qty)


class OrderBook:
    """Python order book using Numba-compiled dictionaries"""
    
    def __init__(self):
        # Use Numba typed dictionaries for JIT compilation
        self.live_orders = NumbaDict.empty(
            key_type=numba.uint64,
            value_type=numba.types.Tuple([numba.uint32, numba.uint32, numba.uint8])
        )
        self.bid_levels = NumbaDict.empty(
            key_type=numba.uint32,
            value_type=numba.uint32
        )
        self.ask_levels = NumbaDict.empty(
            key_type=numba.uint32,
            value_type=numba.uint32
        )
    
    def add(self, oid, side, px, qty):
        add_order(self.live_orders, self.bid_levels, self.ask_levels, oid, side, px, qty)
    
    def cancel(self, oid):
        cancel_order(self.live_orders, self.bid_levels, self.ask_levels, oid)
    
    def execute(self, oid, qty_exec):
        execute_order(self.live_orders, self.bid_levels, self.ask_levels, oid, qty_exec)
    
    def replace(self, oid, new_oid, new_px, new_qty):
        replace_order(self.live_orders, self.bid_levels, self.ask_levels, oid, new_oid, new_px, new_qty)


def main():
    if len(sys.argv) != 2:
        print("Usage: python lob_replay.py <bx_YYYYMMDD.parquet>")
        sys.exit(1)
    
    # 1. Load Parquet file with PyArrow (zero-copy)
    print("Loading Parquet file...")
    table = pq.read_table(sys.argv[1])
    rows = len(table)
    
    # 2. Extract column arrays (zero-copy access to Arrow data)
    ts_arr = table.column('ts').to_numpy()
    oid_arr = table.column('oid').to_numpy()
    side_arr = table.column('side').to_numpy()
    px_arr = table.column('px').to_numpy()
    qty_arr = table.column('qty').to_numpy()
    m_arr = table.column('m').to_pandas()  # String array
    stock_arr = table.column('stock').to_pandas()  # String array
    new_oid_arr = table.column('new_oid').to_numpy()
    new_px_arr = table.column('new_px').to_numpy()
    new_qty_arr = table.column('new_qty').to_numpy()
    
    # 3. Initialize per-symbol order books
    books = {}
    latencies = []
    
    print(f"Processing {rows:,} messages...")
    
    # 4. Execute order book replay with per-message latency tracking
    wall_t0 = time.perf_counter()
    
    for i in range(rows):
        tic = time.perf_counter_ns()
        
        oid = oid_arr[i]
        side = side_arr[i]
        px = px_arr[i]
        qty = qty_arr[i]
        m_type = m_arr.iloc[i]
        symbol = stock_arr.iloc[i]
        
        # Get or create order book for symbol
        if symbol not in books:
            books[symbol] = OrderBook()
        ob = books[symbol]
        
        # Process message
        if m_type == 'A':
            ob.add(oid, side, px, qty)
        elif m_type == 'C':
            ob.cancel(oid)
        elif m_type == 'E':
            ob.execute(oid, qty)
        elif m_type == 'U':
            new_oid = new_oid_arr[i]
            new_px = new_px_arr[i]
            new_qty = new_qty_arr[i]
            ob.replace(oid, new_oid, new_px, new_qty)
        
        toc = time.perf_counter_ns()
        latencies.append(toc - tic)
    
    wall_t1 = time.perf_counter()
    
    # 5. Calculate and display performance metrics
    wall_seconds = wall_t1 - wall_t0
    throughput = rows / wall_seconds
    
    # Sort latencies for percentile calculation
    latencies.sort()
    
    def percentile(p):
        return latencies[int(p * len(latencies))]
    
    # Convert nanoseconds to approximate cycles (assuming 3.2 GHz M1)
    # This is for comparison with C++ version
    ns_to_cycles = 3.2  # Approximate conversion for M1
    
    print("\nLOB Replay Metrics")
    print("──────────────────")
    print(f"Rows processed     : {rows:,}")
    print(f"Total wall time (s): {wall_seconds:.6f}")
    print(f"Throughput (msg/s) : {throughput/1e6:.2f} M")
    print(f"Latency (ns) — p50 : {percentile(0.50):,}")
    print(f"Latency (ns) — p95 : {percentile(0.95):,}")
    print(f"Latency (ns) — p99 : {percentile(0.99):,}")
    print(f"Latency (ns) — max : {max(latencies):,}")
    print(f"Latency (cycles) — p50 : {int(percentile(0.50) * ns_to_cycles):,}")
    print(f"Latency (cycles) — p95 : {int(percentile(0.95) * ns_to_cycles):,}")
    print(f"Latency (cycles) — p99 : {int(percentile(0.99) * ns_to_cycles):,}")
    print(f"Latency (cycles) — max : {int(max(latencies) * ns_to_cycles):,}")
    
    print(f"\nOrder books created: {len(books)}")
    print(f"Symbols processed: {', '.join(list(books.keys())[:10])}" + 
          (f" (and {len(books)-10} more)" if len(books) > 10 else ""))


if __name__ == "__main__":
    main()